DeepSpeech.py

```diff
@@ -1150,6 +1167,33 @@ def _next_epoch(self):
         # Indicates, if there were 'new' epoch(s) provided
         result = False
 
+        # Check if the flag is enabled to decide for early stopping
+        if self._dev:
+            #Make sure that early stop is enabled and validation part is enabled
+            if FLAGS.early_stop is True and FLAGS.validation_step > 0:
+
+                #Check for early stopping condition
+                if(len(self._loss) >= FLAGS.earlystop_nsteps):
+
+                    # Calculate the mean of losses for past epochs
+                    mean_loss = np.mean(self._loss[-FLAGS.earlystop_nsteps:-2])
+                    # Calculate the standard deviation for losses from validation part in the past epochs
+                    std_loss = np.std(self._loss[-FLAGS.earlystop_nsteps:-2])
+                    # Update the list of losses incurred
+                    self._loss = self._loss[-FLAGS.earlystop_nsteps:]
+                    log_debug('Current validation loss: %f, std_of_loss(n_step) :%f mean_loss(n_step): %f' % (self._loss[-1], std_loss, mean_loss))
+
+                    #Making sure slight fluctuations don't bother the early stopping from performing
+                    if self._loss[-1] > np.max(self._loss[:-2]) or (abs(self._loss[-1] - mean_loss) < FLAGS.estop_mean_thresh and std_loss < FLAGS.estop_std_thresh):
+                        #Time to early stop
+                        log_info('EarlyStopp triggered!!')
+            else:
+                #Need to reset the loss from past nsteps
+                self._loss = []
```

Comment:
If early stop is not enabled there won't be any point in accumulating the losses same is the case with validation step disabled.